src/openms/source/QC/DBSuitability.cpp

```diff
@@ -189,110 +529,82 @@ namespace OpenMS
-    vector<double> diffs;
-    for (const auto& pep_id : pep_ids)
+    else
     {
-      double diff = getDecoyDiff_(pep_id);
-      if (diff < DBL_MAX)
+      if (!param_.getValue("force").toBool()) // without xcorr, need to force re-ranking
       {
-        diffs.push_back(diff);
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No cross correlation score found at peptide hit. Only Comet search engine is supported for re-ranking. Set 'force' flag to use the default score for this. This may result in undefined behaviour and is not advised."));
       }
-    }
 
-    if (double(diffs.size()) / pep_ids.size() < 0.2)
-    {
-      throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Under 20 % of peptide identifications have two decoy hits. This is not enough for re-ranking. Use the 'no_rerank' flag to still compute a suitability score."));
+      return pep_hit.getScore();
     }
+  }
 
-    // sort the diffs decreasing and get the percentile one
-    UInt index = round(reranking_cutoff_percentile * diffs.size());
-    
-    if (index >= diffs.size())
+  DBSuitability::SuitabilityData DBSuitability::simulateNoReRanking_(const SuitabilityData& data) const
+  {
+    SuitabilityData simulated_data;
+    simulated_data.num_top_db = data.num_top_db - data.num_re_ranked;
+    simulated_data.num_top_novo = data.num_top_novo + data.num_re_ranked;    
+    return simulated_data;
+  }
+
+  double DBSuitability::calculateCorrectionFactor_(const SuitabilityData& data_full, const SuitabilityData& data_sampled, double sampling_rate) const
+  {
+    if (sampling_rate >= 1 || sampling_rate < 0)
     {
-      return *max_element(diffs.begin(), diffs.end());
+      throw Exception::Precondition(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "The sampling rate has to be element of [0,1).");
     }
 
-    nth_element(diffs.begin(), diffs.begin() + index, diffs.end());
+    double db_slope = (int(data_sampled.num_top_db) - int(data_full.num_top_db)) / (sampling_rate - 1);
+    double deNovo_slope = (int(data_sampled.num_top_novo) - int(data_full.num_top_novo)) / (sampling_rate - 1);
 
-    return diffs[index];
+    return -(db_slope) / (deNovo_slope);
   }
 
-  bool DBSuitability::isNovoHit_(const PeptideHit& hit)
+  void DBSuitability::SuitabilityData::setCorrectionFactor(double factor)
   {
-    const set<String>& accessions = hit.extractProteinAccessionsSet();
-    for (const String& acc : accessions)
+    if (num_top_db == 0 || num_top_novo == 0)
     {
-      if (acc.find(Constants::UserParam::CONCAT_PEPTIDE) == String::npos)
-      {
-        return false;
-      }
+      throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No suitability data found. Can't apply correction factor."));
     }
-    return true;
+    corr_factor = factor;
+    num_top_novo_corr = num_top_novo * factor;
+    suitability_corr = num_top_db / (num_top_db + num_top_novo_corr);
```

Comment:
these three member should be private, just to make sure they are not modified 'by hand', but only via this function.
You'd also need three small `double getCorrectionFactor() const { return corr_factor; }` etc functions in the header then.